### PR TITLE
Performance Improvments on Reflection Based Serialization

### PR DESCRIPTION
**Pull request by shmao**
The PR includes several performance optimizations on reflection based serialization (~60% improvement). Some optimizations also improved de-serialization (~20% improvement). 

@mconnew @roncain @zhenlan , can you please review the PR? Thanks.

### CODE REVIEW

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/CollectionDataContract.cs [hunk 1]
```diff
@@ -919,6 +935,146 @@ internal XmlFormatGetOnlyCollectionReaderDelegate XmlFormatGetOnlyCollectionRead
...
+            private delegate IEnumerator CreateGenericDictionaryEnumeratorDelegate(IEnumerator enumerator);
+
+            private CreateGenericDictionaryEnumeratorDelegate _createGenericDictionaryEnumeratorDelegate;
+
+            internal IEnumerator GetEnumeratorForCollection(object obj, out Type enumeratorReturnType)
+            {
+                IEnumerator enumerator = ((IEnumerable)obj).GetEnumerator();
+                if (Kind == CollectionKind.GenericDictionary)
+                {
+
```
Comment (mconnew) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/CollectionDataContract.cs]: Nit:spacing

File: src/System.Runtime.Serialization.Xml/tests/SerializationTypes.cs [hunk 2]
```diff
@@ -2397,6 +2397,15 @@ public class SimpleTypeWihtMoreProperties
         public List<SimpleTypeWihtMoreProperties> SimpleTypeList { get; set; }
     }
 
+    public class SimpleTypeWihtMoreFields
```
Comment (mconnew) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTypes.cs]: Typo: Wiht->With

**Comment by mconnew** [comment]
:shipit:

**Comment by shmao** [comment]
Thanks for reviewing the PR, @mconnew !

**Pull request merged**